pkg/storage/streams/store.go

```diff
@@ -37,3 +42,295 @@ type Store interface {
+		segmentPath := path.Prepend(fmt.Sprintf("s%d", totalSegments))
+		segmentData := io.LimitReader(awareLimitReader, s.segmentSize)
+		segmentMetatdata := identitySlice
+		putMeta, err := s.segments.Put(ctx, segmentPath, segmentData,
+			segmentMetatdata, expiration)
+		if err != nil {
+			return identityMeta, err
+		}
+		lastSegmentSize = putMeta.Size
+		totalSize = totalSize + putMeta.Size
+	}
+
+	totalSegments = totalSegments + 1
+	identitySegmentData := data
+	lastSegmentPath := path.Prepend("l")
+
+	md := streamspb.MetaStreamInfo{
+		NumberOfSegments: totalSegments,
+		SegmentsSize:     s.segmentSize,
+		LastSegmentSize:  lastSegmentSize,
+		MetaData:         metadata,
+	}
+	lastSegmentMetadata, err := proto.Marshal(&md)
+	if err != nil {
+		return identityMeta, err
+	}
+
+	putMeta, err := s.segments.Put(ctx, lastSegmentPath, identitySegmentData,
+		lastSegmentMetadata, expiration)
+	if err != nil {
+		return identityMeta, err
+	}
+	totalSize = totalSize + putMeta.Size
+
+	resultMeta := Meta{
+		Modified:   time.Now(),
+		Expiration: expiration,
+		Size:       totalSize,
+		Data:       lastSegmentMetadata,
+	}
+
+	return resultMeta, nil
+}
+
+// EOFAwareLimitReader holds reader and status of EOF
+type EOFAwareLimitReader struct {
+	reader io.Reader
+	eof    bool
+}
+
+// EOFAwareReader keeps track of the state, has the internal reader reached EOF
+func EOFAwareReader(r io.Reader) *EOFAwareLimitReader {
+	return &EOFAwareLimitReader{reader: r, eof: false}
+}
+
+func (r *EOFAwareLimitReader) Read(p []byte) (n int, err error) {
+	n, err = r.reader.Read(p)
+	if err == io.EOF {
+		r.eof = true
+	}
+	return n, err
+}
+
+func (r *EOFAwareLimitReader) isEOF() bool {
+	return r.eof
+}
+
+func sizeTuple(totalSegments int64, sizeOfSegments int64) (int64, error, int64, error) {
+	sizePerSegment := float64(totalSegments) / float64(sizeOfSegments)
+	stringSegmentsSize := fmt.Sprintf("%f", sizePerSegment)
+	segmentSizeSlice := strings.Split(stringSegmentsSize, ".")
+	perfectSizedSegments, pErr := strconv.ParseInt(segmentSizeSlice[0], 10, 64)
+	lastSegmentSize, lErr := strconv.ParseInt(segmentSizeSlice[1], 10, 64)
+
+	return perfectSizedSegments, pErr, lastSegmentSize, lErr
+}
+
+func (s *streamStore) Get(ctx context.Context, path paths.Path) (
+	rr ranger.RangeCloser, meta Meta, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	// TODO: return a ranger that knows what the overall size is (from l/<path>)
+	// and then returns the appropriate data from segments s0/<path>, s1/<path>,
+	// ..., l/<path>.
+
+	lastRangerCloser, lastSegmentMeta, err := s.segments.Get(ctx, path.Prepend("l"))
+	if err != nil {
+		return nil, Meta{}, err
+	}
+	totalSize := lastSegmentMeta.Size
+
+	newMeta := Meta{
+		Modified:   lastSegmentMeta.Modified,
+		Expiration: lastSegmentMeta.Expiration,
+		Size:       lastSegmentMeta.Size,
+		Data:       lastSegmentMeta.Data,
+	}
+
+	perfectSizedSegments, pErr, lastSegmentSize, lErr := sizeTuple(totalSize, s.segmentSize)
```

Comment:
I don't really understand this `sizeTuple` thing. We already have all the necessary info in `lastSegmentMeta.Meta`, after unmarshalling it.